### PR TITLE
feat: introduce `TableRouteCache` to `PartitionRuleManager`

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your 

Introduce `TableRouteCache` to `PartitionRuleManager` to reduce overhead of deserializing cached bytes into `TableRoute`.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/common/meta/src/cache/container.rs [hunk 1]
```diff
@@ -101,9 +101,15 @@ where
...
     pub async fn get(&self, key: K) -> Result<Option<V>> {
+        metrics::CACHE_CONTAINER_CACHE_GET
+            .with_label_values(&[&self.name])
+            .inc();
         let moved_init = self.initializer.clone();
         let moved_key = key;
         let init = async move {
+            metrics::CACHE_CONTAINER_CACHE_MISS
+                .with_label_values(&[&self.name])
+                .inc();
```
Comment (fengjiachun) [inline src/common/meta/src/cache/container.rs:112]: Perhaps what we need is a timer to track the time spent, as "time" may require more attention.

File: src/partition/src/error.rs [hunk 2]
```diff
@@ -29,6 +29,13 @@ use crate::expr::PartitionExpr;
 #[snafu(visibility(pub))]
 #[stack_trace_debug]
 pub enum Error {
+    #[snafu(display("Failed to find table route for table id {}", table_id))]
```
Comment (fengjiachun) [inline src/partition/src/error.rs:32]:
```suggestion
    #[snafu(display("Failed to find table route: {}", table_id))]
```

File: src/common/meta/src/cache/container.rs [hunk 3]
```diff
@@ -101,9 +101,15 @@ where
 {
     /// Returns a _clone_ of the value corresponding to the key.
     pub async fn get(&self, key: K) -> Result<Option<V>> {
```
Comment (fengjiachun) [inline src/common/meta/src/cache/container.rs:103]: Can we also provide a `batch_get`? But not in this PR. We have already implemented this in Cached KvBackend: https://github.com/GreptimeTeam/greptimedb/blob/89da42dbc12687157e00dbd4efae75ba154381c1/src/catalog/src/kvbackend/client.rs#L170

**Review by fengjiachun — APPROVED** [review]

File: src/common/meta/src/cache/container.rs (hunk 3 above)
Comment (WenyXu) [reply src/common/meta/src/cache/container.rs:103]: It's a little tricky, we need to employ a lock to ensure consistency.

**Review by waynexia — COMMENTED** [review]
The changes looks good to me. But I'm not sure about other underlying logics (like who should use the cache but not, or who should invalid the cache but not). PTAL @fengjiachun

**Comment by fengjiachun** [comment]
> The changes looks good to me. But I'm not sure about other underlying logics (like who should use the cache but not, or who should invalid the cache but not). PTAL @fengjiachun

I'm o_O ing

**Review by fengjiachun — APPROVED** [review]

**Pull request merged**